File: Medium/1921. Eliminate Maximum Number of Monsters/1921. Eliminate Maximum Number of Monsters.py
from typing import List


class Solution:
    def eliminateMaximum(self, dist: List[int], speed: List[int]) -> int:
        n = len(dist)
        time_to_reach = [0] * n

        for i in range(n):
            time_to_reach[i] = (dist[i] + speed[i] - 1) // speed[i]

        time_to_reach.sort()

        eliminated = 0
        time = 0

        for i in range(n):
            if time_to_reach[i] > time:
                eliminated += 1
                time += 1
            else:
                break

        return eliminated

File: Medium/1921. Eliminate Maximum Number of Monsters/1921. Eliminate Maximum Number of Monsters.java
class Solution {
    public int eliminateMaximum(int[] dist, int[] speed) {
        int n = dist.length;
        int[] timeToReach = new int[n];

        for (int i = 0; i < n; i++) {
            timeToReach[i] = (dist[i] + speed[i] - 1) / speed[i];
        }

        Arrays.sort(timeToReach);
        
        int eliminated = 0;
        int time = 0;

        for (int i = 0; i < n; i++) {
            if (timeToReach[i] > time) {
                eliminated += 1;
                time += 1;
            } else {
                break;
            }
        }

        return eliminated;
    }
}
